File: 981-time-based-key-value-store/981-time-based-key-value-store.js

var TimeMap = function() {
    this._values = {};
    this._stamps = {};
};

/** 
 * @param {string} key 
 * @param {string} value 
 * @param {number} timestamp
 * @return {void}
 */
TimeMap.prototype.set = function(key, value, timestamp) {
    if (!this._values.hasOwnProperty(key)) {
        this._values[key] = [];
        this._stamps[key] = [];
    }
    this._values[key].push(value);
    this._stamps[key].push(timestamp);
};

/** 
 * @param {string} key 
 * @param {number} timestamp
 * @return {string}
 */
TimeMap.prototype.get = function(key, timestamp) {
    if (!this._values.hasOwnProperty(key)) {
        return '';
    }
    
    const times = this._stamps[key];
    
    let lo = 0;
    let hi = times.length - 1;
    
    while (lo <= hi) {
        const mid = (lo + hi) >> 1; // integer divide by 2
        
        if (timestamp < times[mid]) {
            hi = mid - 1;
        } else {
            lo = mid + 1;
        }
    }
    
    return (hi === -1) ? '' : this._values[key][hi];
};

/** 
 * Your TimeMap object will be instantiated and called as such:
 * var obj = new TimeMap()
 * obj.set(key,value,timestamp)
 * var param_2 = obj.get(key,timestamp)
 */

File: 981-time-based-key-value-store/981-time-based-key-value-store.py
from bisect import bisect
from functools import cache


class TimeMap:

    def __init__(self):
        self.values: dict[str, list[str]] = {}
        self.stamps: dict[str, list[int]] = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.values:
            self.values[key] = []
            self.stamps[key] = []
        self.values[key].append(value)
        self.stamps[key].append(timestamp)

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.values:
            return ""
        time = bisect(self.stamps[key], timestamp)
        return self.values[key][time - 1] if time else ""


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)